packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -1509,32 +1527,77 @@ export default function makeKernelKeeper(
+            // process of being deleted. These two clauses are
+            // mutually exclusive.
+
+            if (ownerVatID && !terminated) {
               const vatKeeper = provideVatKeeper(ownerVatID);
               const isReachable = vatKeeper.getReachableFlag(kref);
               if (isReachable) {
                 // the reachable count is zero, but the vat doesn't realize it
                 actions.add(`${ownerVatID} dropExport ${kref}`);
               }
               if (recognizable === 0) {
-                // TODO: rethink this
+                // TODO: rethink this assert
                 // assert.equal(isReachable, false, `${kref} is reachable but not recognizable`);
                 actions.add(`${ownerVatID} retireExport ${kref}`);
               }
-            } else if (recognizable === 0) {
-              // unreachable, unrecognizable, orphaned: delete the
-              // empty refcount here, since we can't send a GC
-              // action without an ownerVatID
-              deleteKernelObject(kref);
+            } else if (ownerVatID && terminated) {
+              // When we're slowly deleting a vat, and one of its
+              // exports becomes unreferenced, we obviously must not
+              // send dropExports or retireExports into the dead vat.
+              // We fast-forward the abandonment that slow-deletion
+              // would have done, then treat the object as orphaned.
+
+              const { vatSlot } = getReachableAndVatSlot(ownerVatID, kref);
+              // delete directly, not abandonKernelObjects(), which
+              // would re-submit to maybeFreeKrefs
+              kvStore.delete(ownerKey);
+              kvStore.delete(`${ownerVatID}.c.${kref}`);
+              kvStore.delete(`${ownerVatID}.c.${vatSlot}`);
+              // now fall through to the orphaned case
+              ownerVatID = undefined;
+            }
+
+            // Now handle objects which were orphaned. NOTE: this
+            // includes objects which were owned by a terminated (but
+            // not fully deleted) vat, where `ownerVatID` was cleared
+            // in the last line of that previous clause (the
+            // fall-through case). Don't try to change this `if
+            // (!ownerVatID)` into an `else if`: the two clauses are
+            // *not* mutually-exclusive.
+
+            if (!ownerVatID) {
+              // orphaned and unreachable, so retire it
+              if (recognizable) {
+                // there are importers, tell them about the
+                // retirement. retireKernelObjects will also
+                // deleteKernelObject
+                retireKernelObjects([kref]);
+              } else {
+                // just delete, without scanning for importers (since
+                // we know there are none). It would be safe to call
+                // retireKernelObjects instead, just slower.
+                deleteKernelObject(kref);
+              }
```

Comment:
Attempting to reduce the real estate claimed by this function:
```suggestion
              // orphaned and unreachable, so retire it
              // we can delete directly if it is also unrecognizable,
              // but otherwise must use retireKernelObjects (which
              // precedes deleteKernelObject with a scan for importers
              // to inform via retireImport actions)
              if (!recognizable) {
                deleteKernelObject(kref);
              } else {
                retireKernelObjects([kref]);
              }
```

Reply:
I wordsmithed this a bit and came up with:
```js
              // orphaned and unreachable, so retire it. If the kref
              // is recognizable, then we need retireKernelObjects()
              // to scan for importers and send retireImports (and
              // delete), else we can call deleteKernelObject directly
	      if (recognizable) {
	        retireKernelObjects([kref]);
	      } else {
	        deleteKernelObject(kref);
	      }
```